ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+has to be done manually by the code:
+
+```ruby
+fiber = Fiber.new do
+  i = 0
+  loop do
+    i += 1
+    puts i
+    Fiber.yield
+  end
+end
+
+fiber.resume
+# => 1
+fiber.resume
+# => 2
+fiber.resume
+# => 3
+```
+
+*NOTE: Each `fiber.resume` call blocks the main thread until a Fiber `yield`s back control.*
+
+#### Reactors
+
+Somewhat less popular, but still very viable are libraries that implement a reactor pattern for
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
```

Comment:
As the class count will grow considerably, I wonder if you want a module/namespace one level deeper for client, worker, errors, etc. There are probably a dozen classes that would only make sense in a client module/namespace. But I understand if Ruby doesn't like much sprawl. In almost all SDKs (Python, TypeScript, Go, etc) we have kept it only two deep.

Reply:
Good point 👍 However, I am planning to do this. An it is possible, because Ruby allows nesting modules AND classes:

```ruby
module Temporal
  class Client
    class Connection
    end
  end
end
```

From the namespacing perspective it doesn't matter whether `Client` is a class or a module, you still reference it like `Temporal::Client::Connection`